### PR TITLE
[10.0][MIG] stock_orderpoint_manual_procurement

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Stock Orderpoint Manual Procurement
===================================

This module allows users to manually start procurements from the list of
reordering rules, based on the quantity that is recommended to be procured.

Usage
=====

Go to 'Inventory > Inventory Control > Reordering Rules' and review the
quantity recommended to be procured. You can now start the procurement for a
single or a list of reordering rules.

The recommended quantity to procure is adjusted to the procurement unit of
measure indicated in the reordering rule.

If you want users to be able to change the recommended quantity to procure,
you should assign them to the security group 'Change quantity in manual
procurements from reordering rules', under 'Settings / Users / Users'.

### CODE REVIEW

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from odoo.tools import float_compare, float_round
+
+
+UNIT = dp.get_precision('Product Unit of Measure')
+
+
+class StockWarehouseOrderpoint(models.Model):
+    _inherit = 'stock.warehouse.orderpoint'
+
+    @api.multi
```
Comment (rousseldenis) [inline stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py]: @lreficent Please put fields definition first

https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#models

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    @api.depends("product_min_qty", "product_id", "qty_multiple")
+    def _compute_procure_recommended(self):
+        op_qtys = self.subtract_procurements_from_orderpoints()
+        for op in self:
+            op.procure_recommended_date = op._get_date_planned(
+                datetime.today())
+            procure_recommended_qty = 0.0
+            prods = op.with_context(
+                location=op.location_id.id).product_id.virtual_available
+            if prods is None:
```
Comment (rousseldenis) [inline stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py]: @lreficent As product_id is required on orderpoints and virtual_available is at least == 0.0, this is not necessary

File: stock_orderpoint_manual_procurement/wizards/make_procurement_orderpoint.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        return res
+
+    @api.multi
+    def make_procurement(self):
+        self.ensure_one()
+        res = []
+        for item in self.item_ids:
+            data = item._prepare_procurement()
+            procurement = self.env['procurement.order'].create(data)
+            procurement.run()
```
Comment (rousseldenis) [inline stock_orderpoint_manual_procurement/wizards/make_procurement_orderpoint.py]: @lreficent Procurements are run on creation. This line avoid to have the 'procurement_autorun_defer' context option to be used. Is this the wanted behaviour?

**Review by rousseldenis — CHANGES_REQUESTED** [review]

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py (hunk 2 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py]: you're right

File: stock_orderpoint_manual_procurement/wizards/make_procurement_orderpoint.py (hunk 3 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/wizards/make_procurement_orderpoint.py]: good point, removing it.

**Comment by LoisRForgeFlow** [comment]
@rousseldenis Thanks for the review, it's been attended.

File: stock_orderpoint_manual_procurement/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    "summary": "Allows to create procurement orders from orderpoints instead "
+               "of relying only on the scheduler.",
+    "version": "10.0.1.0.0",
+    "author": "Eficent, "
+              "Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
+    "category": "Warehouse Management",
+    "depends": [
+        "stock",
+        "stock_orderpoint_uom",
```
Comment (rousseldenis) [inline stock_orderpoint_manual_procurement/__manifest__.py]: @lreficent Why depending on this ? I see the few lines in computing procure_recommended_quantity, but I don't think both should be linked automatically.

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+from odoo.tools import float_compare, float_round
+
+
+UNIT = dp.get_precision('Product Unit of Measure')
+
+
+class StockWarehouseOrderpoint(models.Model):
+    _inherit = 'stock.warehouse.orderpoint'
+
+    procure_recommended_qty = fields.Float(
```
Comment (rousseldenis) [inline stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py:18]: @lreficent Maybe this could be externalized in a more generic module with values that help users to manage orderpoints (such as virtual_available quantity or ...) ?

But that is out of scope of migration process.

File: stock_orderpoint_manual_procurement/__manifest__.py (hunk 4 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/__manifest__.py]: @rousseldenis well, maybe you are right, there is no need to use both modules together. I'll take care of this.

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py (hunk 5 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py:18]: This is also interesting, the need for this field and not for the manual procurement mechanism could exist. I'll look into this too

**Comment by LoisRForgeFlow** [comment]
As a result of https://github.com/OCA/stock-logistics-warehouse/pull/360#pullrequestreview-85798730 -> Glue module to mantain previouse behavior here: https://github.com/OCA/stock-logistics-warehouse/pull/362

**Comment by LoisRForgeFlow** [comment]
Pending:
* [x] : fix tests

File: stock_orderpoint_manual_procurement/__manifest__.py (hunk 4 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/__manifest__.py]: This has been attended

File: stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py (hunk 5 above)
Comment (LoisRForgeFlow) [reply stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py:18]: Hi @rousseldenis 

Rethinking about this, I wonder what would be the use of these field by themselves, I mean: What would be the point of having `procure_recommended_qty` visible if you cannot procure manually?

Comment (rousseldenis) [reply stock_orderpoint_manual_procurement/models/stock_warehouse_orderpoint.py:18]: @lreficent Yes, that's ok. I was thinking even if user should see the next reordering quantities that will be ordered (in any case).

Let behaviour like this

**Comment by rousseldenis** [comment]
@lreficent Needs rebase ?

**Comment by LoisRForgeFlow** [comment]
@rousseldenis seems good now!

**Review by rousseldenis — APPROVED** [review]

**Review by jgrandguillaume — APPROVED** [review]
LGTM

**Review by MiquelRForgeFlow — APPROVED** [review]
